File: dsa-problems/Trees/solutions/python/TRE-015-shuttle-bst-kth-smallest-range.py
import sys

# Increase recursion depth
sys.setrecursionlimit(200000)

class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def insert(node, val):
    if node is None:
        return TreeNode(val)
    if val < node.val:
        node.left = insert(node.left, val)
    else:
        node.right = insert(node.right, val)
    return node

def kth_in_range(values: list[int], L: int, R: int, k: int) -> int:
    root = None
    for v in values:
        root = insert(root, v)
        
    count = 0
    result = -1
    
    def inorder(node):
        nonlocal count, result
        if node is None or result != -1:
            return
            
        # Visit left if potentially valid
        if node.val > L:
            inorder(node.left)
            
        if result != -1: return
        
        if L <= node.val <= R:
            count += 1
            if count == k:
                result = node.val
                return
                
        # Visit right if potentially valid
        if node.val < R:
            inorder(node.right)

    inorder(root)
    return result

def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    idx = 0
    n = int(data[idx]); idx += 1
    values = [int(data[idx + i]) for i in range(n)]
    idx += n
    L = int(data[idx]); idx += 1
    R = int(data[idx]); idx += 1
    k = int(data[idx]) if idx < len(data) else 1
    
    print(kth_in_range(values, L, R, k))

if __name__ == "__main__":
    main()

File: dsa-problems/Trees/solutions/java/TRE-015-shuttle-bst-kth-smallest-range.java
import java.util.*;

class Solution {
    static class TreeNode {
        long val;
        TreeNode left, right;
        TreeNode(long val) { this.val = val; }
    }

    private int count;
    private long result;

    public long kthInRange(long[] values, long L, long R, int k) {
        TreeNode root = null;
        for (long v : values) {
            root = insert(root, v);
        }

        count = 0;
        result = -1;
        inorder(root, L, R, k);
        return result;
    }

    private TreeNode insert(TreeNode node, long val) {
        if (node == null) return new TreeNode(val);
        if (val < node.val) {
            node.left = insert(node.left, val);
        } else {
            node.right = insert(node.right, val);
        }
        return node;
    }

    private void inorder(TreeNode node, long L, long R, int k) {
        if (node == null || result != -1) return;

        // Pruning: if node.val < L, left subtree is useless (all < L)
        // Correct logic: Visit left if it COULD contain values >= L.
        // Left child values are < node.val.
        // If node.val <= L, then left child < L. Useless.
        // So visit left only if node.val > L.
        
        if (node.val > L) {
            inorder(node.left, L, R, k);
        }

        if (result != -1) return; // Check again after returning from left

        if (node.val >= L && node.val <= R) {
            count++;
            if (count == k) {
                result = node.val;
                return;
            }
        }

        // Pruning: Visit right only if node.val < R.
        // If node.val >= R, right child > R. Useless.
        if (node.val < R) {
            inorder(node.right, L, R, k);
        }
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        long[] values = new long[n];
        for (int i = 0; i < n; i++) values[i] = sc.nextLong();
        long L = 0, R = 0;
        if (sc.hasNextLong()) L = sc.nextLong();
        if (sc.hasNextLong()) R = sc.nextLong();
        int k = 0;
        if (sc.hasNextInt()) k = sc.nextInt();

        Solution solution = new Solution();
        System.out.println(solution.kthInRange(values, L, R, k));
        sc.close();
    }
}
